src/package/compressed.rs

```diff
@@ -47,76 +47,112 @@ impl Package {
     /// This intentionally uses a [`BTreeMap`] to ensure that the list of files is sorted
     /// lexicographically. This ensures a reproducible output.
     pub fn create(mut manifest: Manifest, files: BTreeMap<PathBuf, Bytes>) -> miette::Result<Self> {
+        // Create a new conforming manifest if the edition is unknown
         if manifest.edition == Edition::Unknown {
-            manifest = Manifest::new(manifest.package, manifest.dependencies);
+            manifest = Manifest::new(manifest.package.clone(), manifest.dependencies.clone());
         }
 
+        // Ensure the manifest has a package declaration
         if manifest.package.is_none() {
             return Err(miette!(
-                "failed to create package, manifest doesnt contain a package declaration"
+                "failed to create package, manifest doesn't contain a package declaration"
             ));
         }
 
         let mut archive = tar::Builder::new(Vec::new());
 
+        // Add original and resolved manifests
+        Self::add_manifest_to_archive(
+            &mut archive,
+            &manifest.for_publishing()?,
+            Some(MANIFEST_PREFIX),
+            MANIFEST_FILE,
+        )?;
+
+        Self::add_manifest_to_archive(
+            &mut archive,
+            &manifest,
+            None,
+            &format!("{MANIFEST_FILE}.orig"),
+        )?;
+
+        // Add files to the archive
+        for (name, contents) in &files {
+            let mut header = tar::Header::new_gnu();
+            header.set_mode(0o444);
+            header.set_size(contents.len() as u64);
+            archive
+                .append_data(&mut header, name, &contents[..])
+                .into_diagnostic()
+                .wrap_err(miette!("failed to add proto {name:?} to release tar"))?;
+        }
+
+        // Finalize tarball
+        let tar = archive
+            .into_inner()
+            .into_diagnostic()
+            .wrap_err(miette!("failed to assemble tar package"))?;
+
+        // Compress tarball
+        let tgz = Self::compress_tarball(tar)?;
+
+        Ok(Self { manifest, tgz })
+    }
+
+    /// Helper to add a manifest (original or resolved) to the tarball.
+    fn add_manifest_to_archive(
+        archive: &mut tar::Builder<Vec<u8>>,
+        manifest: &Manifest,
+        prefix: Option<&str>,
```

Comment:
Please work with path slices here; instead of pushing strs

Reply:
That way you can also skip the prefix arg as it can be done by concating path slices on the call site

Reply:
I instead introduced a trait `PublishableManifest` and the `ResolvedManifest` newtype. Is that ok for you, too? f77d2faf2d04a4765aace473cea58f2ed5a09cd0